### PR TITLE
[SNOW-748118] Remove delivery guarantee config param

### PR DESCRIPTION
**Pull request by sfc-gh-rcheng**
Streaming has exactly once delivery hard coded
Snowpipe does not support exactly once delivery anymore

Thus remove the config param and corresponding code/logic

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 1]
```diff
@@ -452,14 +448,6 @@ static ConfigDef newConfigDef() {
             JMX_OPT_DEFAULT,
             ConfigDef.Importance.HIGH,
             "Whether to enable JMX MBeans for custom SF metrics")
-        .define(
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:455]: removed from config

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 2]
```diff
@@ -651,11 +636,8 @@ private void insert(final SinkRecord record) {
       }
       // only get offset token once when service context is initialized
       // ignore ingested files
-      // if ingestionDeliveryGuarantee is AT_LEAST_ONCE, ignore offsetPersistedInSnowflake
-      // else discard the record if the record offset is smaller or equal to server side offset
-      if ((ingestionDeliveryGuarantee
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:656]: snowpipe doesnt support at least once anymore. remove this check since it is always false and in an OR

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 3]
```diff
@@ -815,18 +797,13 @@ private long getOffset() {
...
-        } else {
-          this.offsetPersistedInSnowflake.set(Long.parseLong(offsetToken));
-        }
+      ingestionService.ingestFilesWithClientInfo(fileNamesCopy, this.clientSequencer.get());
+      String offsetToken = ingestionService.getClientStatus().getOffsetToken();
+      // Update server side offset
+      if (offsetToken == null) {
+        this.offsetPersistedInSnowflake.set(-1);
       } else {
-        ingestionService.ingestFiles(fileNamesCopy);
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:829]: removed this else because delivery is always exactly once, so it should not fall into this else statement

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 4]
```diff
@@ -815,18 +797,13 @@ private long getOffset() {
...
-        ingestionService.ingestFilesWithClientInfo(fileNamesCopy, this.clientSequencer.get());
-        String offsetToken = ingestionService.getClientStatus().getOffsetToken();
-        // Update server side offset
-        if (offsetToken == null) {
-          this.offsetPersistedInSnowflake.set(-1);
-        } else {
-          this.offsetPersistedInSnowflake.set(Long.parseLong(offsetToken));
-        }
+      ingestionService.ingestFilesWithClientInfo(fileNamesCopy, this.clientSequencer.get());
+      String offsetToken = ingestionService.getClientStatus().getOffsetToken();
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:801]: we wont be calling this API anymore, we should remove offsetPersistedInSnowflake field too.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 5]
```diff
@@ -815,18 +797,13 @@ private long getOffset() {
...
+      // Update server side offset
+      if (offsetToken == null) {
+        this.offsetPersistedInSnowflake.set(-1);
       } else {
-        ingestionService.ingestFiles(fileNamesCopy);
+        this.offsetPersistedInSnowflake.set(Long.parseLong(offsetToken));
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:806]: lets remove both if else.

**Review by sfc-gh-japatel — CHANGES_REQUESTED** [review]
Suggested few changes. 

can you also remove all IT and end to end tests (from test_verify) associated with exactly once? 

Thanks for driving this. 

CC: @sfc-gh-xhuang

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 5 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java]: removed

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 4 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java]: just curious - why are we not calling this API anymore? and why is the offsetPersistedInSnowflake field not necessary? I'm not entirely sure what the importance of this field is/was

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 6]
```diff
@@ -455,11 +440,7 @@ private void init(long recordOffset) {
       // recover will only check pipe status and create pipe if it does not exist.
       recover(pipeCreation);
 
-      // when exactly_once is enabled,fetch clientSequencer and offsetPersistedInSnowflake
-      if (ingestionDeliveryGuarantee
-          == SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE) {
-        initClientInfoForExactlyOnceDelivery();
-      }
+      initClientInfo();
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:443]: the whole idea behind snowpipe's exactly once api was to use streaming channel like persistence layer behind a snowpipe. 
TLDR: Store offset tokens in pipe's metadata using invisible channel object on server side. (Not putting a lot of information here since this is public repo)

Lets remove all occurrences of clientInfo and calling clientConfigure API. We will not need it.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 7]
```diff
@@ -473,30 +454,9 @@ private void init(long recordOffset) {
      * Initialize the client info (clientSequencer and offsetPersistedInSnowflake) by calling
      * ingestion service API configureClient and getClientStatus
      */
-    private void initClientInfoForExactlyOnceDelivery() {
+    private void initClientInfo() {
       ConfigureClientResponse configureClientResponse = ingestionService.configureClient();
       this.clientSequencer.set(configureClientResponse.getClientSequencer());
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:459]: we will not need this anymore. We will solely rely on Snowpipe's basic API which doesnt pass in client info.

**Review by sfc-gh-japatel — COMMENTED** [review]
there are few SF connectors in test/rest_request_template folder pertaining to atleast once and exactly once, can we also remove that? That will reduce end to end running time for few minutes!

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 6 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:443]: Lets discuss this during our 1:1 tomorrow, I dont have enough context to make these changes currently

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java [hunk 8]
```diff
@@ -770,126 +770,6 @@ public void testMissingPropertiesForStreamingClient() {
     }
   }
 
-  /* Service start -> Insert -> Close. service start -> fetch the offsetToken, compare and ingest check data */
-
-  @Test
-  public void testStreamingIngestionWithExactlyOnceSemanticsNoOverlappingOffsets()
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:776]: this is a streaming ingest test!

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java [hunk 9]
```diff
@@ -770,126 +770,6 @@ public void testMissingPropertiesForStreamingClient() {
...
-
-    assert service2.getOffset(new TopicPartition(topic, partition)) == offset + 1;
-
-    service2.closeAll();
-  }
-
-  /* Service start -> Insert -> Close. service start -> fetch the offsetToken, compare and ingest check data */
-
-  @Test
-  public void testStreamingIngestionWithExactlyOnceSemanticsOverlappingOffsets() throws Exception {
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:837]: same as above. lets keep it back.

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java (hunk 8 above)
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:776]: should we keep some of the e2e tests as well? or are those situations redundant with streaming's tests?

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java (hunk 9 above)
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java:837]: reverted, thanks for the catch

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java (hunk 8 above)
Comment (sfc-gh-japatel) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java]: Snowpipe + Exactly once is totally independent from snowpipe streaming. 
There are exactly once tests for snowpipe streaming because we support that. 

We will no longer support snowpipe with exactly once. 

End to end tests which I was suggesting earlier was all ingestionmethod = snowpipe and delivery guarantee = exactly once

We still want to keep snowpipe streaming tests with exactly once (there is no atleast once in snowpipe streaming)

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 10]
```diff
@@ -839,6 +764,8 @@ private long getOffset() {
           name ->
               pipeStatus.updateCommitLag(currentTime - FileNameUtils.fileNameToTimeIngested(name)));
 
+      ingestionService.ingestFiles(fileNamesCopy);
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:767]: lets put this above line 758

We want to update committedOffset only after ingestFiles API have succeeded. 

(This file is the only file which could have potential bugs, lets make sure we revert back to the same changes before exactly once change was introduced)

File: test/test_verify.py [hunk 11]
```diff
@@ -588,12 +530,6 @@ def runTestSet(driver, testSet, nameSalt, enable_stress_test):
             errorExit("Unknown testSet option {}, please input confluent, apache or clean".format(testSet))
 
         execution(testSet, testSuitList1, testCleanEnableList1, testSuitEnableList1, driver, nameSalt)
-        ############################ round 1 ############################
-
-        print("Enable Delivery Guarantee tests:" + str(driver.enableDeliveryGuaranteeTests))
```
Comment (sfc-gh-japatel) [inline test/test_verify.py:593]: awesome! less runtime!

**Comment by sfc-gh-japatel** [comment]
Overall lgtm but will wait for one change about committedOffset and insertFile call

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java (hunk 10 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:767]: moved the line.

good point, i just did a diff against the 1.6.9 tag, and excluding the log and partitionbuffer changes, the only major change is the 'shouldSkipNullValue' method.

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java (hunk 8 above)
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2IT.java]: sweet, thanks for clearing that up

**Comment by sfc-gh-rcheng** [comment]
> Thanks @sfc-gh-rcheng ! I believe there're changes in the Ingest SDK are also related to exactly once, could you help with removing them as well?

@sfc-gh-japatel and i had some concerns over if customers are relying on that API? Should we do a release with a warning about deprecating those api calls, or can we just remove them?

**Comment by sfc-gh-japatel** [comment]
> > Thanks @sfc-gh-rcheng ! I believe there're changes in the Ingest SDK are also related to exactly once, could you help with removing them as well?
> 
> @sfc-gh-japatel and i had some concerns over if customers are relying on that API? Should we do a release with a warning about deprecating those api calls, or can we just remove them?

We can find usage of this APIs and based on usage we can just remove it since it was just PrPr. But Xin might have better suggestions.

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java [hunk 12]
```diff
@@ -651,12 +596,8 @@ private void insert(final SinkRecord record) {
       }
       // only get offset token once when service context is initialized
       // ignore ingested files
-      // if ingestionDeliveryGuarantee is AT_LEAST_ONCE, ignore offsetPersistedInSnowflake
-      // else discard the record if the record offset is smaller or equal to server side offset
-      if ((ingestionDeliveryGuarantee
-                  == SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE
-              || record.kafkaOffset() > this.offsetPersistedInSnowflake.get())
-          && record.kafkaOffset() > processedOffset.get()) {
+      // discard the record if the record offset is smaller or equal to server side offset
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java:599]: nit: this comment in no longer valid

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm, please remove the comment. Otherwise LGTM.

**Comment by sfc-gh-rcheng** [comment]
Pretty big changes in master, merged as well as I could. Please take a look to double check I'm not removing anything important!

**Review by sfc-gh-tjones — APPROVED** [review]

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm

File: src/main/java/com/snowflake/kafka/connector/internal/SnowflakeErrors.java [hunk 13]
```diff
@@ -189,9 +189,6 @@ public enum SnowflakeErrors {
       "Failed to create pipe",
       "User may have insufficient privileges. If this persists, please "
           + "contact Snowflake support. "),
-  ERROR_3006("3006", "Failed to configure client status", "Exception reported by Ingest SDK"),
-  ERROR_3007("3007", "Failed to get client status", "Exception reported by Ingest SDK"),
-  ERROR_3008("3008", "Failed to ingest file with client info", "Exception reported by Ingest SDK"),
```
[the remark this answers is not in the document]
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/SnowflakeErrors.java:194]: added back in, thanks!
